src/glyf.cc

```diff
@@ -344,6 +414,88 @@ bool OpenTypeGLYF::Parse(const uint8_t *data, size_t length) {
   return true;
 }
 
+bool OpenTypeGLYF::TraverseComponentsCountingPoints(
+    Buffer &glyph,
+    uint16_t base_glyph_id,
+    uint32_t level,
+    ComponentPointCount* component_point_count) {
+
+  int16_t num_contours;
+  if (!glyph.ReadS16(&num_contours) ||
+      !glyph.Skip(8)) {
+    return Error("Can't read glyph header.");
+  }
+
+  if (num_contours <= -2) {
+    return Error("Bad number of contours %d in glyph.", num_contours);
+  }
+
+  if (num_contours == 0)
+    return true;
+
+  // FontTools counts a component level for each traversed recursion. We start
+  // counting at level 0. If we reach a level that's deeper than
+  // maxComponentDepth, we expand maxComponentDepth unless it's larger than
+  // the maximum possible depth.
+  if (level > std::numeric_limits<uint16_t>::max()) {
+    return Error("Illegal component depth exceeding 0xFFFF in base glyph id %d.",
+                 base_glyph_id);
+  } else if (level > this->maxp->max_c_depth) {
+    this->maxp->max_c_depth = level;
+    Warning("Component depth exceeds maxp maxComponentDepth "
+            "in glyph %d, adjust limit to %d.",
+            base_glyph_id, level);
+  }
+
+  if (num_contours > 0) {
+    uint16_t num_points = 0;
+    for (int i = 0; i < num_contours; ++i) {
+      // Simple glyph, add contour points.
+      uint16_t tmp_index = 0;
+      if (!glyph.ReadU16(&tmp_index)) {
+        return Error("Can't read contour index %d", i);
+      }
+      num_points = tmp_index + 1;
+    }
+
+    component_point_count->accumulated_component_points += num_points;
+    return true;
+  } else  {
+    assert(num_contours == -1);
```

Comment:
Since we are reading `num_contours` from input data, I don’t think assert is good here. See the checks for `num_contours` in `OpenTypeGLYF::Parse()` (again, I feel uneasy about the code duplication).

Reply:
All previous conditions, `num_contours <= -2`, `0`, `> 0` have been checked, that's why I was using the `assert` here for clarity. Agree that the duplication would be preferably avoided, I have been thinking about that as well. I believe it would mean some kind of recursive call to a paramaterized Parse(), but Parse() has many side effects such as adding to the iov output buffer, which are hard to encapsulate. Also, the intention of this traversal is really only focusing on a fast points count, not on additional validations such as hinting instruction count or logic checks on xmin, xmax etc. coordinates - hence I was considering it acceptable to repeat, also given that the `glyf` table structure is unlikely to change. 

Alternatively, I have been whether it'd make sense to keep some kind of parallel vector of points count in the iterative top level traversal in parse, and pushing points count for components upwards in that, but this does not easily work out as a glyf can be component to multiple other glyfs and there is no requirement of a topological order, so multiple passes would be required IIUC. Also, the idea of creating copies seemed wasteful, as we're anyway traversing over an existing font blob in memory.

Reply:
> All previous conditions, `num_contours <= -2`, `0`, `> 0` have been checked

That is true for the glyph where `TraverseComponentsCountingPoints()` , but is it true for the glyphs that `TraverseComponentsCountingPoints()` is processing? e.g. suppose `A` is a composite glyph of `A.1` and `A.2`, by the time `TraverseComponentsCountingPoints()` is called on `A.1` or r `A.2` they may or may not have been sanitized by `Parse()`, or am I confusing myself?

Reply:
Bear with me if I missed something - I am trying to understand what your direction. IIUC there are two things to discuss here:

- I am using the assert() only to document that num_contours can only be `-1` at this point, at the same code nesting level, previous if conditions have checked for all other values.
- The other thing which you raise here: Order in which all glyphs get sanitized: `OpenTypeGLYF::Parse` iterates glyph by glyph, in the glyph id order. If the glyph has components, at the end of the loop body, this change here starts traversing the components with the intention of computing adherence to maxComponentDepth and maxComponentPoints and adjusting `maxp` when needed. Going off on this traversal, things not related to this traversal are skipped as much as possible, e.g. xmin/xmax validation as well as hinting instructions related things - as they don't affect points count. But each glyph component itself will eventually be visited by the glyph loop in Parse and checks for xmin, max, and TT instructions will be performed.

Reply:
OK, sorry for the confusion. I missed the check for `num_contours <= -2` in `TraverseComponentsCountingPoints()` and thought only `Parse()` checked for it. I’m still uneasy about the duplicate checks, but I doubt it will be easy to miss if in the future `-2` etc. started to have a meaning as the code will error loudly.